### ISSUE TITLE
Solaris user's home directory must be readable for become_user to work with pipelining enabled

### ISSUE THREAD

**Issue by mavit** [ansible/ansible#17908]
##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

`become`
##### ANSIBLE VERSION

```
ansible 2.2.0.0 (detached HEAD 44faad0593) last updated 2016/10/05 11:58:25 (GMT +100)
  lib/ansible/modules/core: (detached HEAD d66983b43e) last updated 2016/10/05 11:58:45 (GMT +100)
  lib/ansible/modules/extras: (detached HEAD 35132b892f) last updated 2016/10/05 11:58:51 (GMT +100)
  configured module search path = Default w/o overrides
```
##### OS / ENVIRONMENT

Solaris.
##### SUMMARY

Directive `become_user` fails when the sudo user can't read the home directory of the ssh user.
##### STEPS TO REPRODUCE

``` bash
chmod og-rx ~
```

``` yaml
- hosts: localhost
  tasks:
    - command: id
      become: true
      become_user: any_user_that_is_not_root
```
##### EXPECTED RESULTS

The command should be run as the `become_user` user.  This worked fine in 2.1.
##### ACTUAL RESULTS

```
TASK [command] *****************************************************************
task path: /home/mavit/etc/ansible/test.yml:3
Using module file /home/mavit/src/ansible/lib/ansible/modules/core/commands/command.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: mavit
<127.0.0.1> EXEC /bin/sh -c '/usr/local/bin/sudo -H -S  -p "[sudo via ansible, key=oxgugeqsazjveqbghcjlfkljckxfnhub] password: " -u any_user_that_is_not_root /bin/sh -c '"'"'echo BECOME-SUCCESS-oxgugeqsazjveqbghcjlfkljckxfnhub; /usr/bin/python'"'"' && sleep 0'
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "<stdin>", line 10, in <module>
  File "/usr/lib/python2.7/posixpath.py", line 371, in abspath
    cwd = os.getcwd()
OSError: [Errno 13] Permission denied

fatal: [localhost]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "invocation": {
        "module_name": "command"
    }, 
    "module_stderr": "Traceback (most recent call last):\n  File \"<stdin>\", line 10, in <module>\n  File \"/usr/lib/python2.7/posixpath.py\", line 371, in abspath\n    cwd = os.getcwd()\nOSError: [Errno 13] Permission denied\n", 
    "module_stdout": "", 
    "msg": "MODULE FAILURE"
}
```

**Comment by bcoca** [comment]
There is not much we can do in this case, but you can set the remote temp to a shared dir (normally /tmp or /var/tmp).

**Comment by mavit** [comment]
Setting `remote_tmp = /tmp` in `ansible.cfg` doesn't help.

I'm probably naïve to some horrible corner cases, but the following hack seems to work for me:

``` diff
diff --git a/lib/ansible/playbook/play_context.py b/lib/ansible/playbook/play_context.py
index b038862..9d39430 100644
--- a/lib/ansible/playbook/play_context.py
+++ b/lib/ansible/playbook/play_context.py
@@ -472,7 +472,7 @@ class PlayContext(Base):
             becomecmd   = None
             randbits    = ''.join(random.choice(string.ascii_lowercase) for x in range(32))
             success_key = 'BECOME-SUCCESS-%s' % randbits
-            success_cmd = pipes.quote('echo %s; %s' % (success_key, cmd))
+            success_cmd = pipes.quote('cd; echo %s; %s' % (success_key, cmd))

             if executable:
                 command = '%s -c %s' % (executable, success_cmd)
```

I should add that this failure is particular to when pipelining is enabled.  If I disable pipelining I hit a different failure, for which I'll log another bug.

**Comment by bcoca** [comment]
soo .. `cd;` should take you to your home dir, which does not make much sense on why you get the error in the first place as that should already be your CWD.

I can only suppose that:
 a) sudo on Solaris requires a readable CWD (no clue why) 
 b) your login is not putting you into your home directory by default

Can you confirm?

**Comment by mavit** [comment]
My understanding of what was happening here was loosely:
1. `sshd` accepts an incoming connection.
2. `sshd` switches user to `remote_user`, changes directory to `~remote_user/`, and starts a shell.
3. The shell starts `sudo`.
4. `sudo` switches user to `become_user`, does _not_ change directory, and starts `python`.
5. `python` tries to inspect its CWD, but fails because it lacks permission for `~remote_user/`.

**Comment by abadger** [comment]
I'm not able to reproduce this:

```
[pts/6@roan /srv/ansible/vanilla]$ chmod og-rx ~                                              *[44faad0]  (11:50:12)
[pts/6@roan /srv/ansible/vanilla]$ ansible --version                                          *[44faad0]  (11:50:34)
ansible 2.2.0.0 (detached HEAD 44faad0593) last updated 2016/10/24 11:48:44 (GMT -700)
  lib/ansible/modules/core: (detached HEAD d66983b43e) last updated 2016/10/24 11:49:42 (GMT -700)
  lib/ansible/modules/extras: (detached HEAD 35132b892f) last updated 2016/10/24 11:49:42 (GMT -700)
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
[pts/6@roan /srv/ansible/vanilla]$ ANSIBLE_NOCOWS=1 ANSIBLE_SSH_PIPELINING=1 ansible-playbook /srv/ansible/troubleshooting/17908.yml --sudo -K -vvv
Using /etc/ansible/ansible.cfg as config file
SUDO password: 
 [WARNING]: provided hosts list is empty, only localhost is available


PLAYBOOK: 17908.yml ************************************************************
1 plays in /srv/ansible/troubleshooting/17908.yml

PLAY [localhost] ***************************************************************

TASK [setup] *******************************************************************
Using module file /srv/ansible/vanilla/lib/ansible/modules/core/system/setup.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: badger
<127.0.0.1> EXEC /bin/sh -c 'sudo -H -S  -p "[sudo via ansible, key=ubrpfttewjpytgwcsnfvfogeurcidtof] password: " -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-ubrpfttewjpytgwcsnfvfogeurcidtof; /usr/bin/python'"'"' && sleep 0'
ok: [localhost]

TASK [command] *****************************************************************
task path: /srv/ansible/troubleshooting/17908.yml:3
Using module file /srv/ansible/vanilla/lib/ansible/modules/core/commands/command.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: badger
<127.0.0.1> EXEC /bin/sh -c 'sudo -H -S  -p "[sudo via ansible, key=fqzptlcnxzrnqizfgrjoasazbntoyamr] password: " -u nobody /bin/sh -c '"'"'echo BECOME-SUCCESS-fqzptlcnxzrnqizfgrjoasazbntoyamr; /usr/bin/python'"'"' && sleep 0'
changed: [localhost] => {
    "changed": true, 
    "cmd": [
        "id"
    ], 
    "delta": "0:00:00.002430", 
    "end": "2016-10-24 11:50:48.626095", 
    "invocation": {
        "module_args": {
            "_raw_params": "id", 
            "_uses_shell": false, 
            "chdir": null, 
            "creates": null, 
            "executable": null, 
            "removes": null, 
            "warn": true
        }, 
        "module_name": "command"
    }, 
    "rc": 0, 
    "start": "2016-10-24 11:50:48.623665", 
    "stderr": "", 
    "stdout": "uid=99(nobody) gid=99(nobody) groups=99(nobody) context=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023", 
    "stdout_lines": [
        "uid=99(nobody) gid=99(nobody) groups=99(nobody) context=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023"
    ], 
    "warnings": []
}

PLAY RECAP *********************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0
[pts/6@roan ~]$ cat /srv/ansible/troubleshooting/17908.yml                                                (11:52:23)
- hosts: localhost
  tasks:
    - command: id
      become: true
      become_user: nobody
```

IS there anything else that you're doing differently?

**Comment by abadger** [comment]
needs_info

**Comment by ansibot** [comment]
@mavit ping, this issue is still waiting on your feedback. We will close the issue if you do not respond.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@mavit ping, this issue is still waiting on your feedback. We will close the issue if you do not respond.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by abadger** [comment]
I'm going to guess that we might need Solaris in order to reproduce this.

**Comment by mavit** [comment]
You can run [Solaris](http://www.oracle.com/technetwork/server-storage/solaris11/downloads/index.html) for free "[for the purpose of developing, testing, prototyping and demonstrating your applications, and not for any other purpose](http://www.oracle.com/technetwork/licenses/solaris-cluster-express-license-167852.html)".  Problematically, this does not include security updates -- for that you need a support contract.  

Something like [OpenIndiana](https://www.openindiana.org/) may well be close enough for this kind of testing, but I haven't tried it myself.

**Comment by perolausson** [comment]
I think I have a good guess as to what went on here.

Terms: 

become_user = the non-root user the play wants to become
remote_user = the user ansible connects to target with

The current logic (maybe it has changed?) in Ansible is:

if the home directory of remote_user is not executable by become_user
then 
become become_user in the **home directory of the become_user**
else
become become_user in the **home directory of the remote_user**

This appears to me to be undocumented and causes interesting problems and non-deterministic behaviour.

The following test playbook exemplifies this behaviour (tested on Ubuntu 17.04/Ansible 2.4.0.0)

```
---
- hosts: ubuntu
  tasks:
    - name: Protect home directory
      file:
        path: ~/
        mode: "{{mode}}"
        state: directory

    - name: Create test user
      user:
        name: zzbecome
        state: present
      become: true

    - name: Become someone else
      shell: id && pwd
      become: true
      become_user: zzbecome
      register: result

    - debug: msg="{{result.stdout_lines}}"

    - name: Delete test user
      user:
        name: zzbecome
        state: absent
        remove: yes
      become: true
```

When run with "ansible-playbook test_become.yml -e mode=755" the result is:

```
TASK [debug] **************************************************************************************************************************************************************************
ok: [ubuntu] => {
    "msg": [
        "uid=1002(zzbecome) gid=1002(zzbecome) groups=1002(zzbecome)", 
        "/home/ansible"
    ]
}
```
When run with "ansible-playbook test_become.yml -e mode=700": 

```
TASK [debug] **************************************************************************************************************************************************************************
ok: [ubuntu] => {
    "msg": [
        "uid=1002(zzbecome) gid=1002(zzbecome) groups=1002(zzbecome)", 
        "/home/zzbecome"
    ]
}
```

When run with "ansible-playbook test_become.yml -e mode=744": 

```
TASK [Become someone else] ************************************************************************************************************************************************************
fatal: [ubuntu]: FAILED! => {"changed": false, "failed": true, "module_stderr": "Shared connection to ubuntu closed.\r\n", "module_stdout": "\r\nTraceback (most recent call last):\r\n  File \"/tmp/ansible_5RvsCO/ansible_module_command.py\", line 213, in <module>\r\n    main()\r\n  File \"/tmp/ansible_5RvsCO/ansible_module_command.py\", line 185, in main\r\n    rc, out, err = module.run_command(args, executable=executable, use_unsafe_shell=shell, encoding=None, data=stdin)\r\n  File \"/tmp/ansible_5RvsCO/ansible_modlib.zip/ansible/module_utils/basic.py\", line 2813, in run_command\r\nOSError: [Errno 13] Permission denied: '/home/ansible'\r\n", "msg": "MODULE FAILURE", "rc": 1}
        to retry, use: --limit @/home/pao/Development/Automation/test_become.retry
```

**Comment by elnappo** [comment]
I had the same problem, my home directory was a mounted NFS share (I know, wtf?!) with `root squash` set...
Setting `remote_tmp = /tmp` in `ansible.cfg` resolved the issue for me.

**Comment by sadanand25** [comment]
Hello Team,



Still issue is not resolved ...  please help to get it resolved

The full traceback is:
  File "/tmp/ansible_gm8qvinh/ansible_modlib.zip/ansible/module_utils/basic.py", line 2762, in run_command
    cmd = subprocess.Popen(args, **kwargs)
  File "/opt/lib/python3.5/subprocess.py", line 947, in __init__
    restore_signals, start_new_session)
  File "/opt/lib/python3.5/subprocess.py", line 1551, in _execute_child
    raise child_exception_type(errno_num, err_msg)

fatal: [XXX]: FAILED! => {
    "changed": false,
    "cmd": "/usr/bin/mksysb -X -e -G -p -a -i /imagebkp/XXX",
    "invocation": {
        "module_args": {
            "backup_crypt_files": true,
            "backup_dmapi_fs": true,
            "create_map_files": false,
            "exclude_files": true,
            "exclude_wpar_files": true,
            "extended_attrs": true,
            "name": "XXX",
            "new_image_data": true,
            "software_packing": false,
            "storage_path": "/imagebkp",
            "use_snapshot": false
        }
    },
    "msg": "[Errno 13] The file access permissions do not allow the specified action.",
    "rc": 13
}

**Comment by mator** [comment]
@sadanand25 we're talking about solaris issue here, probably resolved. Please tell me why do you reporting upon solaris when running on AIX (mksysb) ? Thanks.

**Comment by Jiehong** [comment]
I seem to face the same issue on suse linux here.

**Comment by jwbernin** [comment]
I'm encountering this same issue on RHEL 8 when running Ansible as a user whose home directory is shared via NFS.

Ansible version:
```$ ansible --version
ansible 2.8.0
  config file = /nfshome/jberning/Projects/homelab/ansible.cfg
  configured module search path = ['/nfshome/jberning/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.6/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 3.6.8 (default, Apr  3 2019, 17:26:03) [GCC 8.2.1 20180905 (Red Hat 8.2.1-3)]
```

**Comment by bcoca** [comment]
@jwbernin  read https://github.com/ansible/ansible/issues/17908#issuecomment-338923412 for a solution, the NFS issue is different cause, this ticket is for a particular issue with solaris (which I still am unable to reproduce in Openindiana).

**Comment by mator** [comment]
git version of ansible
```
$ ansible --version
ansible 2.10.0.dev0
```

solaris 10 x86 (VM):
```
[root@solaris10 /# ls -ld ~oracle
drwxr-xr-x   2 oracle   oinstall       2 Jan 22  2013 /export/home/oracle

ansiblehost$ ansible -a "id" solaris10 -b --become-user=oracle
solaris10 | CHANGED | rc=0 >>
uid=54321(oracle) gid=54321(oinstall)

solaris10# su - oracle
Oracle Corporation      SunOS 5.10      Generic Patch   January 2005
solaris10$ pwd
/export/home/oracle
solaris10$ ls -ld .
drwxr-xr-x   2 oracle   oinstall       2 Jan 22  solaris102013 .
$ bash
bash-3.2$ chmod og-rx ~
bash-3.2$ ls -ld .
drwx------   2 oracle   oinstall       2 Jan 22  2013 .

ansiblehost$ ansible -a "id" solaris10 -b --become-user=oracle
solaris10 | CHANGED | rc=0 >>
uid=54321(oracle) gid=54321(oinstall)
```

solaris 11 sparc:
```
ansiblehost$ ansible -a "id" solaris11sparc -b --become-user=user1
solaris11sparc | FAILED | rc=-1 >>
Failed to set permissions on the temporary files Ansible needs to create when becoming an unprivileged user (rc: 2, err: chown: /var/tmp/ansible-tmp-1578320747.89-65410834799024/: Not owner
chown: /var/tmp/ansible-tmp-1578320747.89-65410834799024/AnsiballZ_command.py: Not owner
}). For information on working around this, see https://docs.ansible.com/ansible/become.html#becoming-an-unprivileged-user

root@solaris11:/# su - user1
Oracle Corporation      SunOS 5.11      11.4    Aug 2018
user1@solaris11:~$ ls -ld .
drwxr-xr-x   2 user1    staff          8 Jan  6 14:32 .
user1@solaris11:~$ id -a
uid=102(user1) gid=10(staff) groups=10(staff)
```

**Comment by ansibot** [comment]
@mavit This issue is waiting for your response. Please respond or the issue will be closed.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@mavit You have not responded to information requests in this issue so we will assume it no longer affects you. If you are still interested in this, please create a new issue with the requested information.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)
